src/libraries/Common/src/System/Net/IPv4AddressHelper.Common.cs

```diff
@@ -105,32 +101,44 @@ internal static unsafe bool IsValid(char* name, int start, ref int end, bool all
         //                 / "2" %x30-34 DIGIT     ; 200-249
         //                 / "25" %x30-35          ; 250-255
         //
-        internal static unsafe bool IsValidCanonical(char* name, int start, ref int end, bool allowIPv6, bool notImplicitFile)
+        internal static bool IsValidCanonical<TChar>(ReadOnlySpan<TChar> name, out int bytesConsumed, bool allowIPv6, bool notImplicitFile)
+            where TChar : unmanaged, IBinaryInteger<TChar>
         {
             int dots = 0;
             int number = 0;
             bool haveNumber = false;
             bool firstCharIsZero = false;
+            int current;
+
+            Debug.Assert(typeof(TChar) == typeof(char) || typeof(TChar) == typeof(byte));
 
-            while (start < end)
+            bytesConsumed = 0;
+            for (current = 0; current < name.Length; current++)
             {
-                char ch = name[start];
+                TChar ch = name[current];
+
                 if (allowIPv6)
                 {
-                    // for ipv4 inside ipv6 the terminator is either ScopeId, prefix or ipv6 terminator
-                    if (ch == ']' || ch == '/' || ch == '%')
+                    // For an IPv4 address nested inside an IPv6, the terminator is either ScopeId ('%'), prefix ('/') or ipv6 address terminator (']')
+                    if (ch == TChar.CreateChecked('%') || ch == TChar.CreateChecked('/') || ch == TChar.CreateChecked(']'))
+                    {
                         break;
+                    }
                 }
-                else if (ch == '/' || ch == '\\' || (notImplicitFile && (ch == ':' || ch == '?' || ch == '#')))
+                // For a normal IPv4 address, the terminator is the prefix ('/' or its counterpart, '\'). If notImplicitFile is set, the terminator
+                // is one of the characters which signify the start of the rest of the URI - the port number (':'), query string ('?') or fragment ('#')
+                else if (ch == TChar.CreateChecked('/') || ch == TChar.CreateChecked('\\')
+                    || (notImplicitFile && (ch == TChar.CreateChecked(':') || ch == TChar.CreateChecked('?') || ch == TChar.CreateChecked('#'))))
                 {
                     break;
                 }
 
-                if (char.IsAsciiDigit(ch))
+                if (IPAddressParser.TryParseInteger(IPAddressParser.Decimal, ch, out int parsedCharacter))
                 {
-                    if (!haveNumber && (ch == '0'))
+                    // A number starting with zero should be interpreted in base 8 / octal
+                    if (!haveNumber && ch == TChar.CreateChecked('0'))
                     {
-                        if ((start + 1 < end) && name[start + 1] == '0')
+                        if (current + 1 < name.Length && name[current + 1] == TChar.CreateChecked('0'))
```

Comment:
(applies to many other places)

Switching from a `char*` to a `ReadOnlySpan<TChar>` will introduce extra bounds checks in many places like this. Some of those could be avoided by using tricks like this:
```c#
if ((uint)(current + 1) < (uint)name.Length && name[current + 1] == TChar.CreateTruncating('0'))
```

Reply:
I agree. I've checked, and as far as I can tell the IP address parsing already performs pre-emptive bounds checks in a JIT-visible manner.

I don't think the impact of this switch from `char*` will be significant though: other methods called by `Uri.CheckAuthorityHelper` already use `ReadOnlySpan<char>`. Doing so also eliminates a few places in `Uri` which would pin the underlying string, so it improves GC efficiency slightly.